[front] - feat(obs): ingest search output

JulesBelveze opened the pull request:
## Description

This PR implements analytics tracking for agent tool outputs by ingesting search result documents into ES. 

When agents execute search tools, the documents retrieved are now stored in an analytics index to enable observability around data source usage and retrieval patterns. The feature is gated behind the `agent_tool_outputs_analytics` feature flag.
## Risk

Low behind FF

## Tests

- [x] Locally

## Deploy Plan

- Apply the Elasticsearch alias migration in both EU and US regions (`20251219_create_agent_tool_outputs_alias.http`)
- Deploy front

JulesBelveze committed:
[analytics/indices] - feature: rename and update analytics index mappings

 - Renamed retrieval output index files to tool output index files to better reflect their purpose
 - Added a new field 'mcp_server_name' to the tool outputs mapping for enhanced server identification

JulesBelveze committed:
[analytics/migrations] - feature: create new alias for agent tool outputs index

 - Remove old alias for front.agent_tool_outputs* indices
 - Create a new write index alias for front.agent_tool_outputs with front.agent_tool_outputs_1 as the target index

JulesBelveze committed:
[front/types/assistant] - feature: track agent retrieval output analytics

 - Added a new interface for analytics data related to agent retrieval outputs
 - The new structure will capture detailed information including message, workspace, conversation IDs, and data source details for analytical purposes

JulesBelveze committed:
[front/lib/api] - feature: introduce Elasticsearch alias for agent tool outputs

 - Added a constant for the Elasticsearch alias related to agent tool outputs
 - Included the mapping for the new agent tool outputs indices directory

JulesBelveze committed:
[front] - feature: add analytics for agent tool outputs

 - Enable storage of agent tool outputs in Elasticsearch for analytical purposes
 - Feature limited to 'dust_only' stage for initial deployment and testing

JulesBelveze committed:
[analytics_queue] - feature: enhance analytics with agent tool outputs

 - Implement analytics tracking for data retrieved by agent tool outputs
 - Add feature flag control for new analytics capability
 - Store analytics data for search result types in Elasticsearch
 - Refactor analytics activities to include data source information in analytics
 - Improve analytics activities with access to various new resources and models

JulesBelveze committed:
[sdks] - feature: add analytics for agent tool outputs

 - Include 'agent_tool_outputs_analytics' in the WhitelistableFeaturesSchema to enable tracking and analytics for agent tool outputs.

JulesBelveze reviewed (commented):
@flvndvd I don't have better than `agent_retrieval_outputs` 🤷🏼

JulesBelveze committed:
[analytics] - refactor: rename agent_tool_outputs to agent_document_outputs

 - Update index and alias names, replacing 'tool' with 'document' to better reflect their content
 - Modify Elasticsearch related code to use the new AGENT_DOCUMENT_OUTPUTS alias
 - Remove old migrations related to agent_tool_outputs and add new ones for agent_document_outputs

JulesBelveze committed:
[front/lib/resources] - feature: add resource for MCP server configuration management

 - Implement AgentMCPServerConfigurationResource class to handle fetching and deletion of server configurations
 - Enable fetching configurations by server IDs for a specific workspace
 - Provide a method to delete a configuration transactionally within a workspace context

JulesBelveze committed:
[front/lib/resources] - refactor: centralize fetchOutputItemsByActionIds logic

 - Extracted repeated fetchOutputItemsByActionIds logic into a separate static method in AgentMCPActionResource
 - Replaced inline querying of output items in `enrichActionsWithOutputItems` method with a call to the new centralized method

[front/temporal/analytics_queue] - refactor: utilize resources for querying server configurations

 - Replaced direct Model queries with resource-based methods for fetching MCP server configurations
 - Simplified transformation of server configuration IDs to sIDs by leveraging the new resource methods in analytical activities
 - Updated extractRetrievalOutputs to use the new method in AgentMCPActionResource for fetching output items by action IDs

JulesBelveze committed:
[front/lib/resources] - refactor: rename fetchBySIds to fetchByStringIds and parse IDs

 - Function 'fetchBySIds' renamed to more descriptive 'fetchByStringIds' which clearly reflects its purpose
 - String IDs are now converted to numeric before querying, filtering out any non-numeric IDs

[front/temporal/analytics_queue] - refactor: update analytics activities to use new ID handling

 - 'collectToolUsageFromMessage' and 'extractRetrievalOutputs' activities updated to use the renamed and refactored 'fetchByStringIds' method
 - Maps now use numeric IDs converted to strings for consistency with new function behavior

JulesBelveze committed:
[front/lib/resources] - refactor: update fetching of server configurations by ID

- Replaced `fetchByStringIds` with `fetchByModelIds` to directly use numeric ModelIds for queries
- Added `fetchByModelIdsAsStrings` to handle conversion from string IDs to numeric ModelIds

[front/temporal/analytics_queue] - fix: adjust activities to use new server config fetching method

- Modified `collectToolUsageFromMessage` and `extractRetrievalOutputs` to use `fetchByModelIdsAsStrings` following refactoring in resource file

flvndvd commented on front/lib/resources/agent_mcp_server_configuration_resource.ts:
```diff
@@ -0,0 +1,79 @@
+    return configs.map((c) => new this(this.model, c.get()));
+  }
+
+  static async fetchByModelIdsAsStrings(
```
🤮 Can we cast outside of the resource and directly use `fetchByModelIds`?

JulesBelveze committed:
[front] - refactor: optimize retrieval output extraction for analytics

 - Combined two asynchronous data fetches into one Promise.all to reduce waiting time
 - Streamlined the structure of AgentRetrievalOutputAnalyticsData and reduced unnecessary loops
 - Added logging to warn about missing data source IDs or when data source views are not found during retrieval output processing

JulesBelveze committed:
[front/lib/resources] - refactor: remove fetchByModelIdsAsStrings method

 - Eliminated an unnecessary method converting string IDs to integers for fetching configurations
 - Simplified the codebase by removing an additional step in fetching MCP server configurations

JulesBelveze committed:
[front/lib/resources] - feature: add documentation for MCP Server Configuration resource

- Explain the architectural decision of keeping MCP Server Configuration as a standalone resource
- Clarify the resource's role in complexity management, independent access, separation of concerns, lifecycle management, and following established codebase patterns

[front/temporal] - refactor: update analytics activities to use ModelIds for server configurations

- Change method names to better reflect their purpose in extracting retrieval documents
- Introduce hashing for generated document ids to ensure compatibility with Elasticsearch
- Utilize numeric ModelIds for fetching MCP server configurations, improving consistency and type-safety in analytics tracking code

JulesBelveze commented on front/lib/resources/agent_mcp_server_configuration_resource.ts:
```diff
@@ -0,0 +1,79 @@
+  extends ReadonlyAttributesType<AgentMCPServerConfigurationModel> {}
+
+// eslint-disable-next-line @typescript-eslint/no-unsafe-declaration-merging
+export class AgentMCPServerConfigurationResource extends BaseResource<AgentMCPServerConfigurationModel> {
```
Can definitely do 😇

JulesBelveze committed:
fix: lint/format
